plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,501 @@
+
+  ## [REQUIRED] The user to authenticate as, must have admin rights
+  username = "admin"
+
+  ## [REQUIRED] The user's password to authenticate with
+  password = "Passw0rd"
+`
+
+func (o *OpenStack) SampleConfig() string {
+	return sampleConfig
+}
+
+func init() {
+	o := OpenStack{
+		Domain: "Default",
+	}
+	inputs.Add("openstack", func() telegraf.Input { return &o })
+}
+
+func (o *OpenStack) Gather(acc telegraf.Accumulator) error {
+
+	// Authenticate against Keystone and get a token provider
+	authOptions := gophercloud.AuthOptions{
+		IdentityEndpoint: o.IdentityEndpoint,
+		DomainName:       o.Domain,
+		TenantName:       o.Project,
+		Username:         o.Username,
+		Password:         o.Password,
+	}
+
+	provider, err := openstack.AuthenticatedClient(authOptions)
+	if err != nil {
+		return fmt.Errorf("Unable to authenticate OpenStack user: %v", err)
+	}
+
+	// Gather resources
+	// Don't bomb out here, some data is better than none, the 'gather'
+	// functions will check for validity before continuing
+	projectMap, err := getProjectMap(provider)
+	if err != nil {
+		log.Println("W! failed to get projects: " + err.Error())
+	}
+	hypervisorList, err := getHypervisorList(provider)
+	if err != nil {
+		log.Println("W! failed to get hypervisors: " + err.Error())
+	}
+	flavorMap, err := getFlavorMap(provider)
+	if err != nil {
+		log.Println("W! failed to get flavors: " + err.Error())
+	}
+	serverList, err := getServerList(provider)
+	if err != nil {
+		log.Println("W! failed to get servers: " + err.Error())
+	}
+	volumeList, err := getVolumeList(provider)
+	if err != nil {
+		log.Println("W! failed to get volumes: " + err.Error())
+	}
+	storagePoolList, err := getStoragePools(provider)
+	if err != nil {
+		log.Println("W! failed to get storage pools: " + err.Error())
+	}
+
+	// Calculate statistics
+	gatherIdentityStatistics(acc, projectMap)
+	gatherHypervisorStatistics(acc, hypervisorList)
+	gatherServerStatistics(acc, projectMap, flavorMap, serverList)
+	gatherVolumeStatistics(acc, projectMap, volumeList)
+	gatherStoragePoolStatistics(acc, storagePoolList)
+
+	return nil
+}
+
+func getProjectMap(provider *gophercloud.ProviderClient) (ProjectMap, error) {
+
+	identity, err := openstack.NewIdentityV3(provider, gophercloud.EndpointOpts{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to create V3 identity client: %v", err)
+	}
+
+	page, err := projects.List(identity, &projects.ListOpts{}).AllPages()
+	if err != nil {
+		return nil, fmt.Errorf("unable to list projects: %v", err)
+	}
+
+	projectList, err := projects.ExtractProjects(page)
+	if err != nil {
+		return nil, fmt.Errorf("unable to extract projects: %v", err)
+	}
+
+	projectMap := ProjectMap{}
+	for _, project := range projectList {
+		projectMap[project.ID] = project
+	}
+
+	return projectMap, nil
+}
+
+func getHypervisorList(provider *gophercloud.ProviderClient) (HypervisorList, error) {
+
+	compute, err := openstack.NewComputeV2(provider, gophercloud.EndpointOpts{})
```

Comment:
Maybe we should create one compute client and pass it into these functions?  We could store it across calls to Gather too.  It might make sense to store the other clients as well.